injector/injector.go

```diff
@@ -0,0 +1,302 @@
+	nonLinearStartTag              = "<NonLinear>"
+	nonLinearEndTag                = "</NonLinear>"
+	videoClickStartTag             = "<VideoClicks>"
+	videoClickEndTag               = "</VideoClicks>"
+	trackingEventStartTag          = "<TrackingEvents>"
+	trackingEventEndTag            = "</TrackingEvents>"
+	clickTrackingStartTag          = "<ClickTracking><![CDATA["
+	clickTrackingEndTag            = "]]></ClickTracking>"
+	impressionStartTag             = "<Impression><![CDATA["
+	impressionEndTag               = "]]></Impression>"
+	errorStartTag                  = "<Error><![CDATA["
+	errorEndTag                    = "]]></Error>"
+	nonLinearClickTrackingStartTag = "<NonLinearClickTracking><![CDATA["
+	nonLinearClickTrackingEndTag   = "]]></NonLinearClickTracking>"
+	companionClickThroughStartTag  = "<CompanionClickThrough><![CDATA["
+	companionClickThroughEndTag    = "]]></CompanionClickThrough>"
+	tracking                       = "tracking"
+	companionclickthrough          = "companionclickthrough"
+	nonlinearclicktracking         = "nonlinearclicktracking"
+	impression                     = "impression"
+	err                            = "error"
+	clicktracking                  = "clicktracking"
+)
+
+type Injector interface {
+	InjectTracker(vastXML string, NURL string) string
+}
+
+type VASTEvents struct {
+	Errors                  []string
+	Impressions             []string
+	VideoClicks             []string
+	NonLinearClickTracking  []string
+	CompanionClickThrough   []string
+	LinearTrackingEvents    map[string][]string
+	NonLinearTrackingEvents map[string][]string
+	CompanionTrackingEvents map[string][]string
+}
+
+type TrackerInjector struct {
+	replacer macros.Replacer
+	events   VASTEvents
+	me       metrics.MetricsEngine
+	provider *macros.MacroProvider
+}
+
+var trimRunes = "\t\r\b\n "
+
+func NewTrackerInjector(replacer macros.Replacer, provider *macros.MacroProvider, events VASTEvents) *TrackerInjector {
+	return &TrackerInjector{
+		replacer: replacer,
+		provider: provider,
+		events:   events,
+	}
+}
+
+func (ti *TrackerInjector) InjectTracker(vastXML string, NURL string) string {
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+
+	injectTracker := false
+	injectVideoClicks := false
+	inlineWrapperTagFound := false
+	wrapperTagFound := false
+	impressionTagFound := false
+	errorTagFound := false
+	creativeId := ""
+	isCreative := false
+	companionTagFound := false
+	nonLinearTagFound := false
+
+	reader := strings.NewReader(vastXML)
+	decoder := xml.NewDecoder(reader)
+
+	for {
+		t, err := decoder.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			switch tt.Name.Local {
```

Comment:
I'm curious about your opinion: I believe the nested switch statement could be easier to read if it's put in its own function. Thoughts?
```
111
112         switch tt := t.(type) {
113         case xml.StartElement:
    +           addXmlStartElement(tt.Name.Local)
114 -            switch tt.Name.Local {
115 -            case "Wrapper":
116 -                wrapperTagFound = true
117 -            case "Creative":
118 -                isCreative = true
119 -                for _, attr := range tt.Attr {
120 -                    if strings.ToLower(attr.Name.Local) == "adid" {
121 -                        creativeId = attr.Value
122 -                    }
123 -                }
124 -            case "Linear":
125 -                injectVideoClicks = true
126 -                injectTracker = true
127 -            case "VideoClicks":
128 -                injectVideoClicks = false
129 -                encoder.Flush()
130 -                encoder.EncodeToken(tt)
131 -                encoder.Flush()
132 -                ti.addClickTrackingEvent(&outputXML, creativeId, false)
133
134 -                continue
135 -            case "NonLinearAds":
136 -                injectTracker = true
137 -            case "TrackingEvents":
138 -                if isCreative {
139 -                    injectTracker = false
140 -                    encoder.Flush()
141 -                    encoder.EncodeToken(tt)
142 -                    encoder.Flush()
143 -                    ti.addTrackingEvent(&outputXML, creativeId, false)
144 -                    continue
145 -                }
146 -            }
injector/injector.go
```

Reply:
Done